src/lib/ReactCore_Internal.re

```diff
@@ -164,97 +164,139 @@ module Make = (Implementation: HostImplementation) => {
+    type instanceUpdate =
+      | InstanceUpdate(
+                        instance('state, 'action, 'elementType),
+                        instance('state, 'action, 'elementType)
+                      ): instanceUpdate;
     type update = {
+      oldId: int,
+      newId: int,
+      instanceUpdate,
+      stateChanged: bool,
+      subTreeChanged: subtreeChange
+    };
+    type switchComponent = {
       oldId: int,
       newId: int,
       oldOpaqueInstance: opaqueInstance,
       newOpaqueInstance: opaqueInstance,
-      componentChanged: bool,
-      stateChanged: bool,
-      subTreeChanged: bool
+      subTreeChanged: subtreeChange
     };
     type entry =
       | UpdateInstance(update)
-      | NewRenderedElement(renderedElement);
+      | SwitchComponent(switchComponent);
     type t = ref(list(entry));
+    type topLevelUpdate = {
+      typ: subtreeChange,
+      updateLog: t
+    };
     let create = () => ref([]);
     let add = (updateLog, x) => updateLog := [x, ...updateLog^];
   };
   module Render = {
-    let createSelf = (~instance) : self(_) => {
-      state: instance.iState,
+    let createSelf = (~state, ~component, ~pendingStateUpdates) : self(_) => {
+      state,
       reduce: (payloadToAction, payload) => {
-        logString("Calling reduce on " ++ instance.component.debugName);
+        logString("Calling reduce on " ++ component.debugName);
         let action = payloadToAction(payload);
-        let stateUpdate = instance.component.reducer(action);
-        instance.pendingStateUpdates :=
-          [stateUpdate, ...instance.pendingStateUpdates^]
+        let stateUpdate = component.reducer(action);
+        pendingStateUpdates := [stateUpdate, ...pendingStateUpdates^];
       },
-      act: (action) => {
-        logString("Calling act on " ++ instance.component.debugName);
-        let stateUpdate = instance.component.reducer(action);
-        instance.pendingStateUpdates :=
-          [stateUpdate, ...instance.pendingStateUpdates^]
+      act: action => {
+        logString("Calling act on " ++ component.debugName);
+        let stateUpdate = component.reducer(action);
+        pendingStateUpdates := [stateUpdate, ...pendingStateUpdates^];
       }
     };
-    let createInstance = (~component, ~element, ~instanceSubTree, ~subElements) => {
-      let iState = component.initialState();
-      let id = GlobalState.instanceIdCounter^;
-      incr(GlobalState.instanceIdCounter);
+    let createInitialSelf = (~component) => {
+      let pendingStateUpdates = ref([]);
+      (
+        createSelf(
+          ~component,
+          ~state=component.initialState(),
+          ~pendingStateUpdates
+        ),
+        pendingStateUpdates
+      );
+    };
+    let createSelf = (~instance) =>
+      createSelf(
+        ~component=instance.component,
+        ~state=instance.iState,
+        ~pendingStateUpdates=instance.pendingStateUpdates
+      );
+    let createInstance =
```

Comment:
Maybe this function is redundant and we should just create a record wherever it's used now?